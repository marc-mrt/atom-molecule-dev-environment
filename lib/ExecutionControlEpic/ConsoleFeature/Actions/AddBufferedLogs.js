"use babel";
// @flow

import type { ConsoleLog } from "../Types/types";

export type SetBufferedLogsForTaskAction = {
  type: "ADD_CONSOLE_LOGS",
  payload: {
    logs: Array<ConsoleLog>,
  },
};

export function addBufferedLogsForTask(
  logs: Array<ConsoleLog>,
): SetBufferedLogsForTaskAction {
  return {
    type: "ADD_CONSOLE_LOGS",
    payload: {
      logs,
    },
  };
}
