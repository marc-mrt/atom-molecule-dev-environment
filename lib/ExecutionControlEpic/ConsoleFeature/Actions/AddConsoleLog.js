"use babel";
// @flow

import type { ConsoleLog } from "../Types/types.js.flow";

export type SetConsoleLogsForTaskAction = {
  type: "ADD_CONSOLE_LOG",
  payload: {
    consoleLog: ConsoleLog,
  },
};

export function addConsoleLogsForTask(
  log: ConsoleLog,
): SetConsoleLogsForTaskAction {
  return {
    type: "ADD_CONSOLE_LOG",
    payload: {
      consoleLog: log,
    },
  };
}
