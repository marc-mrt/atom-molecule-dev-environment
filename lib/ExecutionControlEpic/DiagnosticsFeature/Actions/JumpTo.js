"use babel";
// @flow

import type { Range } from "../../LanguageServerProtocolFeature/Types/standard";
import Uri from "vscode-uri";

export type JumpToAction = () => void;

export function jumpTo(
  path: string,
  range: Range,
  pending: boolean,
): JumpToAction | void {
  if (path.length > 0) {
    let uri = Uri.parse(path);
    return () => {
      return global.atom.workspace.open(
        uri.scheme === "file" ? uri.path : path,
        {
          initialLine: range.start.line ? range.start.line : 0,
          initialColumn: range.start.character ? range.start.character : 0,
          pending,
        },
      );
    };
  }
}
