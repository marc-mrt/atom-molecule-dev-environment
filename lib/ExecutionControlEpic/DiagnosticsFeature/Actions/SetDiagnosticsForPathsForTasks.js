"use babel";
// @flow

import type { MoleculeDiagnostic } from "../Types/types.js.flow";

export type PathDef = {
  uri: string,
  diagnostics: Array<MoleculeDiagnostic>,
};

export type TaskDef = {
  taskId: string,
  paths: Array<PathDef>,
};

export type SetDiagnosticsForPathsForTasksAction = {
  type: "SET_DIAGNOSTICS_FOR_PATHS_FOR_TASKS",
  payload: {
    tasks: Array<TaskDef>,
  },
};

export function setDiagnosticsForPathsForTasks(
  taskDefs: Array<TaskDef>,
): SetDiagnosticsForPathsForTasksAction {
  return {
    type: "SET_DIAGNOSTICS_FOR_PATHS_FOR_TASKS",
    payload: {
      tasks: taskDefs,
    },
  };
}
