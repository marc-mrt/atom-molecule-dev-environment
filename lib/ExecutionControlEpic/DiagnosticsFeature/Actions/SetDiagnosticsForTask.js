"use babel";
// @flow

import type { MoleculeDiagnostic } from "../Types/types.js.flow";

export type PublishDiagnosticsParamsForTask = {
  uri: string,
  task: string,
  diagnostics: Array<MoleculeDiagnostic>,
};

export type SetDiagnosticsForTaskAction = {
  type: "SET_DIAGNOSTICS_FOR_PATH_FOR_TASK",
  payload: PublishDiagnosticsParamsForTask,
};

export function setDiagnosticsForTask(
  publishedDiagnostics: PublishDiagnosticsParamsForTask,
): SetDiagnosticsForTaskAction {
  return {
    type: "SET_DIAGNOSTICS_FOR_PATH_FOR_TASK",
    payload: publishedDiagnostics,
  };
}
