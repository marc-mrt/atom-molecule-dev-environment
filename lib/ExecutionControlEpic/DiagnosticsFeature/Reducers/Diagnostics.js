"use babel";
// @flow

import type {
  MoleculeDiagnostic,
  DiagnosticSeverity,
} from "../Types/types.js.flow";
import { List, Map } from "immutable";

export default function Diagnostics(
  state: DiagnosticsReducer = Map(),
  action: any,
): DiagnosticsReducer {
  switch (action.type) {
    case "SET_DIAGNOSTICS_FOR_PATH_FOR_TASK":
      return action.payload.diagnostics
        .reduce((red, diagnostic) => {
          return red.update(
            diagnostic.path || action.payload.uri,
            (old = List()) => old.push(diagnostic),
          );
        }, Map())
        .entrySeq()
        .reduce((red, entry) => {
          return red.setIn(
            [action.payload.task, entry[0]],
            entry[1].reduce((acc, diagnostic) => {
              return acc.update(diagnostic.severity, (diagnostics = List()) =>
                diagnostics.push(diagnostic),
              );
            }, Map()),
          );
        }, state.setIn([action.payload.task, action.payload.uri], Map()));
    case "SET_DIAGNOSTICS_FOR_PATHS_FOR_TASKS":
      return action.payload.tasks.reduce(
        (red, taskDef) =>
          red.update(taskDef.taskId, currentTaskReducer =>
            taskDef.paths.reduce(
              (acc, pathDef) =>
                acc.set(
                  pathDef.path,
                  pathDef.diagnostics
                    .reduce(
                      (reduction, diagnostic) => reduction.push(diagnostic),
                      List(),
                    )
                    .reduce((reduction, diagnostic) => {
                      return reduction.update(
                        diagnostic.severity,
                        (diagnostics = List()) => diagnostics.push(diagnostic),
                      );
                    }, Map()),
                ),
              currentTaskReducer || Map(),
            ),
          ),
        state,
      );
    case "REMOVE_DIAGNOSTICS_OF_TASK":
      return state.delete(action.payload.task);
    default:
      return state;
  }
}

export type DiagnosticsReducer = Map<
  string,
  Map<string, Map<DiagnosticSeverity, List<MoleculeDiagnostic>>>,
>;
