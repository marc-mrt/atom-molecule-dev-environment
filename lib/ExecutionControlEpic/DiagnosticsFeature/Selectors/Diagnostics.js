"use babel";
// @flow

import type { DiagnosticsReducer } from "../Reducers/Diagnostics";
import type { DiagnosticSeverity, MoleculeDiagnostic } from "../Types/types";
import { List, Map } from "immutable";

export function selectDiagnosticsOfTask(
  diagnostics: DiagnosticsReducer,
  taskId: string,
): Map<string, Map<DiagnosticSeverity, List<MoleculeDiagnostic>>> {
  return diagnostics.get(taskId) || Map();
}

export function selectDiagnosticsOfStep(
  diagnostics: Map<string, Map<DiagnosticSeverity, List<MoleculeDiagnostic>>>,
  step: number,
): Map<string, Map<DiagnosticSeverity, List<MoleculeDiagnostic>>> {
  return diagnostics.map(severityMap =>
    severityMap.map(list => list.filter(d => d.step === step)),
  );
}

export function selectLastDiagnostics(
  diagnostics: Map<string, Map<DiagnosticSeverity, List<MoleculeDiagnostic>>>,
): Map<string, Map<DiagnosticSeverity, List<MoleculeDiagnostic>>> {
  const max = diagnostics
    .map(severityMap =>
      severityMap.reduce((red, cur) => red.concat(cur), List()),
    )
    .reduce((red, cur) => red.concat(cur), List())
    .reduce((red, value) => (value.step > red ? value.step : red), 0);
  return selectDiagnosticsOfStep(diagnostics, max);
}
