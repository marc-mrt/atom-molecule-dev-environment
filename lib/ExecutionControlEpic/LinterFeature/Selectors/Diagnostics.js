"use babel";
// @flow

import type { MoleculeDiagnostic } from "../../DiagnosticsFeature/Types/types";
import type { DiagnosticsReducer } from "../../DiagnosticsFeature/Reducers/Diagnostics";
import { List } from "immutable";

function isValidLocation(diagnostic: MoleculeDiagnostic): boolean {
  return diagnostic.path && diagnostic.path.length > 0;
}

export default function DiagnosticsToLint(
  diagnostics: DiagnosticsReducer,
): List<MoleculeDiagnostic> {
  return diagnostics
    .reduce(
      (acc, map) =>
        map.reduce(
          (list, severities) =>
            list
              .concat(severities.get(1) || List())
              .concat(severities.get(2) || List()),
          acc,
        ),
      List(),
    )
    .filter(isValidLocation);
}
