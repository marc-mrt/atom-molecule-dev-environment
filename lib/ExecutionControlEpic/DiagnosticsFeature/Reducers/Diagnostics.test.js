"use babel";
// @flow

import type { DiagnosticsReducer } from "./Diagnostics";
import Diagnostics from "./Diagnostics";
import type { RemoveDiagnosticsForTaskAction } from "../Actions/RemoveDiagnosticsForTask";
import { removeDiagnosticsForTask } from "../Actions/RemoveDiagnosticsForTask";
import type { SetDiagnosticsForTaskAction } from "../Actions/SetDiagnosticsForTask";
import { setDiagnosticsForTask } from "../Actions/SetDiagnosticsForTask";
import { setDiagnosticsForPathsForTasks } from "../Actions/SetDiagnosticsForPathsForTasks";
import {
  gulpDiagnostics,
  dockerDBInfo1,
  dockerWeb1Error1,
  dockerWeb1Error2,
  dockerWeb2Warning1,
} from "../Fake/Data/FakeDiagnostics";
import { List, Map } from "immutable";

describe("Diagnostics Reducer", () => {
  describe("SET_DIAGNOSTICS_FOR_PATH_FOR_TASK", () => {
    it("should init the Map with the diagnostic", () => {
      let state: DiagnosticsReducer = Map();
      let action: SetDiagnosticsForTaskAction = setDiagnosticsForTask({
        uri: "this is a uri",
        task: gulpDiagnostics[0].task,
        diagnostics: gulpDiagnostics,
      });
      let subject = Diagnostics(state, action);

      expect(subject).toEqual(
        Map([
          [
            gulpDiagnostics[0].task,
            Map([
              [
                "this is a uri",
                Map([[gulpDiagnostics[0].severity, List(gulpDiagnostics)]]),
              ],
            ]),
          ],
        ]),
      );
    });
  });

  describe("SET_DIAGNOSTICS_FOR_PATHS_FOR_TASKS", () => {
    it("should init the Map with the diagnostic", () => {
      let subject = Diagnostics(
        undefined,
        setDiagnosticsForPathsForTasks([
          {
            taskId: "2",
            paths: [
              {
                path: "/test1",
                diagnostics: [],
              },
              {
                path: "/test2",
                diagnostics: [
                  {
                    severity: 1,
                    message: "aaaaah",
                  },
                  {
                    severity: 2,
                    message: "uuuuuh",
                  },
                ],
              },
            ],
          },
          {
            taskId: "3",
            paths: [
              {
                path: "/test3",
                diagnostics: [
                  {
                    severity: 1,
                    message: "test",
                  },
                  {
                    severity: 1,
                    message: "test2",
                  },
                  {
                    severity: 3,
                    message: "oups",
                  },
                ],
              },
            ],
          },
          {
            taskId: "4",
            paths: [
              {
                path: "/test1",
                diagnostics: [
                  {
                    severity: 2,
                    message: ":o",
                  },
                ],
              },
              {
                path: "/test2",
                diagnostics: [
                  {
                    severity: 2,
                    message: ":)",
                  },
                ],
              },
            ],
          },
        ]),
      );

      expect(subject).toMatchSnapshot();
    });
  });

  describe("REMOVE_DIAGNOSTICS_OF_TASK", () => {
    it("should remove every task of given task id", () => {
      let state: DiagnosticsReducer = Map([
        [
          dockerDBInfo1.task,
          Map([
            ["uri", Map([[dockerDBInfo1.severity, List([dockerDBInfo1])]])],
          ]),
        ],
        [
          dockerWeb1Error2.task,
          Map([
            [
              "uri",
              Map([
                [
                  dockerWeb1Error1.severity,
                  List([dockerWeb1Error1, dockerWeb1Error2]),
                ],
              ]),
            ],
          ]),
        ],
        [
          dockerWeb2Warning1.task,
          Map([
            [
              "uri",
              Map([[dockerWeb2Warning1.severity, List([dockerWeb2Warning1])]]),
            ],
          ]),
        ],
      ]);
      [dockerDBInfo1, dockerWeb1Error1, dockerWeb1Error2, dockerWeb2Warning1];
      let action: RemoveDiagnosticsForTaskAction = removeDiagnosticsForTask(
        dockerWeb1Error1.task,
      );
      let subject = Diagnostics(state, action);

      expect(subject).toEqual(
        Map([
          [
            dockerDBInfo1.task,
            Map([
              ["uri", Map([[dockerDBInfo1.severity, List([dockerDBInfo1])]])],
            ]),
          ],
          [
            dockerWeb2Warning1.task,
            Map([
              [
                "uri",
                Map([
                  [dockerWeb2Warning1.severity, List([dockerWeb2Warning1])],
                ]),
              ],
            ]),
          ],
        ]),
      );
    });
  });
});
