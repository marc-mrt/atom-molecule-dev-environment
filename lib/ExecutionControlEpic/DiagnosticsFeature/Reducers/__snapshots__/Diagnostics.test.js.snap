// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diagnostics Reducer SET_DIAGNOSTICS_FOR_PATHS_FOR_TASKS should init the Map with the diagnostic 1`] = `
Immutable.Map {
  "2": Immutable.Map {
    "/test1": Immutable.Map {},
    "/test2": Immutable.Map {
      1: Immutable.List [
        Object {
          "message": "aaaaah",
          "severity": 1,
        },
      ],
      2: Immutable.List [
        Object {
          "message": "uuuuuh",
          "severity": 2,
        },
      ],
    },
  },
  "3": Immutable.Map {
    "/test3": Immutable.Map {
      1: Immutable.List [
        Object {
          "message": "test",
          "severity": 1,
        },
        Object {
          "message": "test2",
          "severity": 1,
        },
      ],
      3: Immutable.List [
        Object {
          "message": "oups",
          "severity": 3,
        },
      ],
    },
  },
  "4": Immutable.Map {
    "/test1": Immutable.Map {
      2: Immutable.List [
        Object {
          "message": ":o",
          "severity": 2,
        },
      ],
    },
    "/test2": Immutable.Map {
      2: Immutable.List [
        Object {
          "message": ":)",
          "severity": 2,
        },
      ],
    },
  },
}
`;
